pitr restore fail for "failed to insert rows into gc_delete_range: [parser:1064]"

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Start log backup
2. Do full backup
3. run sysbench workload
4. execute flashback ddl before step 3
5. pause log backup during flashback
6. resume log backup
7. Sleep sometime, and wait checkpoint move forward
8. Do a PiTR restore, and check data consistency (restored-ts = ts1)

### 2. What did you expect to see? (Required)
pitr restore success
### 3. What did you see instead (Required)
fail
```
[2023/12/11 09:12:15.863 +00:00] [ERROR] [restore.go:64] ["failed to restore"] [error="failed to insert rows into gc_delete_range: [parser:1064]You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 48 near \"\" "] [errorVerbose="[parser:1064]You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 48 near \"\" \ngithub.com/pingcap/errors.AddStack\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/errors.go:174\ngithub.com/pingcap/errors.(*Error).GenWithStackByArgs\n\t/go/pkg/mod/github.com/pingcap/errors@v0.11.5-0.20221009092201-b66cddb77c32/normalize.go:164\ngithub.com/pingcap/tidb/pkg/util.SyntaxError\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/pkg/util/misc.go:159\ngithub.com/pingcap/tidb/pkg/session.(*session).ParseWithParams\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/pkg/session/session.go:1705\ngithub.com/pingcap/tidb/pkg/session.(*session).ExecuteInternal\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/pkg/session/session.go:1594\ngithub.com/pingcap/tidb/br/pkg/gluetidb.(*tidbSession).ExecuteInternal\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/gluetidb/glue.go:181\ngithub.com/pingcap/tidb/br/pkg/restore.(*Client).InsertGCRows\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/restore/client.go:3484\ngithub.com/pingcap/tidb/br/pkg/task.restoreStream\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/task/stream.go:1414\ngithub.com/pingcap/tidb/br/pkg/task.RunStreamRestore\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/task/stream.go:1179\ngithub.com/pingcap/tidb/br/pkg/task.RunRestore\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/pkg/task/restore.go:621\nmain.runRestoreCommand\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/restore.go:63\nmain.newStreamRestoreCommand.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/restore.go:237\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992\nmain.main\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/main.go:58\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:267\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1650\nfailed to insert rows into gc_delete_range"] [stack="main.runRestoreCommand\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/restore.go:64\nmain.newStreamRestoreCommand.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/restore.go:237\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992\nmain.main\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/br/br/cmd/br/main.go:58\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:267"]
```
### 4. What is your TiDB version? (Required)

master

--- a/br/pkg/restore/client.go
+++ b/br/pkg/restore/client.go
@@ -3481,8 +3481,12 @@ func (rc *Client) InsertGCRows(ctx context.Context) error {
 			// (job_id, elem_id, start_key, end_key, ts)
 			paramsList = append(paramsList, newJobID, params.ElemID, params.StartKey, params.EndKey, ts)
 		}
-		if err := rc.db.se.ExecuteInternal(ctx, query.Sql, paramsList...); err != nil {
-			return errors.Trace(err)
+		if len(paramsList) > 0 {
+			// trim the ',' behind the query.Sql if exists
+			sql := strings.TrimSuffix(query.Sql, ",")
+			if err := rc.db.se.ExecuteInternal(ctx, sql, paramsList...); err != nil {
+				return errors.Trace(err)
+			}
 		}
 	}
 	return nil

--- a/br/pkg/restore/client.go
+++ b/br/pkg/restore/client.go
@@ -3483,6 +3483,7 @@ func (rc *Client) InsertGCRows(ctx context.Context) error {
 		}
 		if len(paramsList) > 0 {
 			// trim the ',' behind the query.Sql if exists
+			// that's when the rewrite rule of the last table id is not exist
 			sql := strings.TrimSuffix(query.Sql, ",")
 			if err := rc.db.se.ExecuteInternal(ctx, sql, paramsList...); err != nil {
 				return errors.Trace(err)

--- a/br/pkg/restore/client_test.go
+++ b/br/pkg/restore/client_test.go
@@ -586,6 +586,43 @@ func TestSetSpeedLimit(t *testing.T) {
 	}
 }
 
+var deleteRangeQueryList = []*stream.PreDelRangeQuery{
+	{
+		Sql: "INSERT IGNORE INTO mysql.gc_delete_range VALUES (%?, %?, %?, %?, %?), (%?, %?, %?, %?, %?)",
+		ParamsList: []stream.DelRangeParams{
+			{
+				JobID:    1,
+				ElemID:   1,
+				StartKey: "a",
+				EndKey:   "b",
+			},
+			{
+				JobID:    1,
+				ElemID:   2,
+				StartKey: "b",
+				EndKey:   "c",
+			},
+		},
+	},
+	{
+		// When the last table id has no rewrite rule
+		Sql: "INSERT IGNORE INTO mysql.gc_delete_range VALUES (%?, %?, %?, %?, %?),",
+		ParamsList: []stream.DelRangeParams{
+			{
+				JobID:    2,
+				ElemID:   1,
+				StartKey: "a",
+				EndKey:   "b",
+			},
+		},
+	},
+	{
+		// When all the tables have no rewrite rule
+		Sql:        "INSERT IGNORE INTO mysql.gc_delete_range VALUES ",
+		ParamsList: nil,
+	},
+}
+
 func TestDeleteRangeQuery(t *testing.T) {
 	ctx := context.Background()
 	m := mc
@@ -619,40 +656,63 @@ func TestDeleteRangeQuery(t *testing.T) {
 
 	client.RunGCRowsLoader(ctx)
 
-	client.RecordDeleteRange(&stream.PreDelRangeQuery{
-		Sql: "INSERT IGNORE INTO mysql.gc_delete_range VALUES (%?, %?, %?, %?, %?), (%?, %?, %?, %?, %?)",
-		ParamsList: []stream.DelRangeParams{
-			{
-				JobID:    1,
-				ElemID:   1,
-				StartKey: "a",
-				EndKey:   "b",
-			},
-			{
-				JobID:    1,
-				ElemID:   2,
-				StartKey: "b",
-				EndKey:   "c",
+	for _, query := range deleteRangeQueryList {
+		client.RecordDeleteRange(query)
+	}
+	querys := client.GetGCRows()
+	require.Equal(t, len(querys), len(deleteRangeQueryList))
+	for i, query := range querys {
+		expected_query := deleteRangeQueryList[i]
+		require.Equal(t, expected_query.Sql, query.Sql)
+		require.Equal(t, len(expected_query.ParamsList), len(query.ParamsList))
+		for j := range expected_query.ParamsList {
+			require.Equal(t, expected_query.ParamsList[j], query.ParamsList[j])
+		}
+
+	}
+
+}
+
+func TestDeleteRangeQueryExec(t *testing.T) {
+	ctx := context.Background()
+	m := mc
+	mockStores := []*metapb.Store{
+		{
+			Id: 1,
+			Labels: []*metapb.StoreLabel{
+				{
+					Key:   "engine",
+					Value: "tiflash",
+				},
 			},
 		},
-	})
+		{
+			Id: 2,
+			Labels: []*metapb.StoreLabel{
+				{
+					Key:   "engine",
+					Value: "tiflash",
+				},
+			},
+		},
+	}
 
-	querys := client.GetGCRows()
-	require.Equal(t, len(querys), 1)
-	require.Equal(t, querys[0].Sql, "INSERT IGNORE INTO mysql.gc_delete_range VALUES (%?, %?, %?, %?, %?), (%?, %?, %?, %?, %?)")
-	require.Equal(t, len(querys[0].ParamsList), 2)
-	require.Equal(t, querys[0].ParamsList[0], stream.DelRangeParams{
-		JobID:    1,
-		ElemID:   1,
-		StartKey: "a",
-		EndKey:   "b",
-	})
-	require.Equal(t, querys[0].ParamsList[1], stream.DelRangeParams{
-		JobID:    1,
-		ElemID:   2,
-		StartKey: "b",
-		EndKey:   "c",
-	})
+	g := gluetidb.New()
+	retryCnt := 1
+	client := restore.NewRestoreClient(fakePDClient{
+		stores:     mockStores,
+		retryTimes: &retryCnt,
+	}, nil, defaultKeepaliveCfg, false)
+	err := client.Init(g, m.Storage)
+	require.NoError(t, err)
+
+	client.RunGCRowsLoader(ctx)
+
+	for _, query := range deleteRangeQueryList {
+		client.RecordDeleteRange(query)
+	}
+
+	require.NoError(t, client.InsertGCRows(ctx))
 }
 
 func MockEmptySchemasReplace() *stream.SchemasReplace {

--- a/br/pkg/restore/client_test.go
+++ b/br/pkg/restore/client_test.go
@@ -668,9 +668,7 @@ func TestDeleteRangeQuery(t *testing.T) {
 		for j := range expected_query.ParamsList {
 			require.Equal(t, expected_query.ParamsList[j], query.ParamsList[j])
 		}
-
 	}
-
 }
 
 func TestDeleteRangeQueryExec(t *testing.T) {

--- a/br/pkg/stream/BUILD.bazel
+++ b/br/pkg/stream/BUILD.bazel
@@ -55,7 +55,7 @@ go_test(
     ],
     embed = [":stream"],
     flaky = True,
-    shard_count = 25,
+    shard_count = 26,
     deps = [
         "//br/pkg/storage",
         "//br/pkg/streamhelper",

--- a/br/pkg/stream/rewrite_meta_rawkv.go
+++ b/br/pkg/stream/rewrite_meta_rawkv.go
@@ -748,7 +748,7 @@ func (bdr *brDelRangeExecWrapper) PrepareParamsList(sz int) {
 func (bdr *brDelRangeExecWrapper) RewriteTableID(tableID int64) (int64, bool) {
 	newTableID, exists := bdr.globalTableIdMap[tableID]
 	if !exists {
-		log.Warn("failed to find the downstream id when rewrite delete range", zap.Int64("tableID", newTableID))
+		log.Warn("failed to find the downstream id when rewrite delete range", zap.Int64("old tableID", tableID))
 	}
 	return newTableID, exists
 }

--- a/br/pkg/stream/rewrite_meta_rawkv.go
+++ b/br/pkg/stream/rewrite_meta_rawkv.go
@@ -746,11 +746,11 @@ func (bdr *brDelRangeExecWrapper) PrepareParamsList(sz int) {
 }
 
 func (bdr *brDelRangeExecWrapper) RewriteTableID(tableID int64) (int64, bool) {
-	tableID, exists := bdr.globalTableIdMap[tableID]
+	newTableID, exists := bdr.globalTableIdMap[tableID]
 	if !exists {
-		log.Warn("failed to find the downstream id when rewrite delete range", zap.Int64("tableID", tableID))
+		log.Warn("failed to find the downstream id when rewrite delete range", zap.Int64("tableID", newTableID))
 	}
-	return tableID, exists
+	return newTableID, exists
 }
 
 func (bdr *brDelRangeExecWrapper) AppendParamsList(jobID, elemID int64, startKey, endKey string) {

--- a/br/pkg/stream/rewrite_meta_rawkv_test.go
+++ b/br/pkg/stream/rewrite_meta_rawkv_test.go
@@ -998,6 +998,62 @@ func TestDeleteRangeForMDDLJob(t *testing.T) {
 	require.Equal(t, encodeTableIndexKey(mDDLJobTable1NewID, int64(3)), qargs.ParamsList[0].StartKey)
 }
 
+func TestDeleteRangeForMDDLJob2(t *testing.T) {
+	midr := newMockInsertDeleteRange()
+	partitionMap := map[int64]int64{
+		mDDLJobPartition0OldID: mDDLJobPartition0NewID,
+		mDDLJobPartition1OldID: mDDLJobPartition1NewID,
+		mDDLJobPartition2OldID: mDDLJobPartition2NewID,
+	}
+	tableReplace0 := &TableReplace{
+		TableID:      mDDLJobTable0NewID,
+		PartitionMap: partitionMap,
+	}
+	tableReplace1 := &TableReplace{
+		TableID: mDDLJobTable1NewID,
+	}
+	tableMap := map[int64]*TableReplace{
+		mDDLJobTable0OldID: tableReplace0,
+		mDDLJobTable1OldID: tableReplace1,
+	}
+	dbReplace := &DBReplace{
+		DbID:     mDDLJobDBNewID,
+		TableMap: tableMap,
+	}
+	schemaReplace := MockEmptySchemasReplace(midr, map[int64]*DBReplace{
+		mDDLJobDBOldID: dbReplace,
+	})
+	var qargs *PreDelRangeQuery
+	// drop schema
+	err := schemaReplace.restoreFromHistory(dropSchemaJob)
+	require.NoError(t, err)
+	qargs = <-midr.queryCh
+	require.Equal(t, len(qargs.ParamsList), len(mDDLJobALLNewTableIDSet))
+	for _, params := range qargs.ParamsList {
+		_, exist := mDDLJobALLNewTableKeySet[params.StartKey]
+		require.True(t, exist)
+	}
+	require.Equal(t, "INSERT IGNORE INTO mysql.gc_delete_range VALUES (%?, %?, %?, %?, %?),(%?, %?, %?, %?, %?),(%?, %?, %?, %?, %?),(%?, %?, %?, %?, %?),(%?, %?, %?, %?, %?)", qargs.Sql)
+
+	// drop schema - lose rewrite rule of table 1
+	tableMap_incomplete := map[int64]*TableReplace{
+		mDDLJobTable0OldID: tableReplace0,
+	}
+	dbReplace.TableMap = tableMap_incomplete
+	schemaReplace = MockEmptySchemasReplace(midr, map[int64]*DBReplace{
+		mDDLJobDBOldID: dbReplace,
+	})
+	err = schemaReplace.restoreFromHistory(dropSchemaJob)
+	require.NoError(t, err)
+	qargs = <-midr.queryCh
+	require.Equal(t, len(qargs.ParamsList), len(mDDLJobALLNewPartitionIDSet)+1)
+	for _, params := range qargs.ParamsList {
+		_, exist := mDDLJobALLNewTableKeySet[params.StartKey]
+		require.True(t, exist)
+	}
+	require.Equal(t, "INSERT IGNORE INTO mysql.gc_delete_range VALUES (%?, %?, %?, %?, %?),(%?, %?, %?, %?, %?),(%?, %?, %?, %?, %?),(%?, %?, %?, %?, %?),", qargs.Sql)
+}
+
 func TestCompatibleAlert(t *testing.T) {
 	require.Equal(t, ddl.BRInsertDeleteRangeSQLPrefix, `INSERT IGNORE INTO mysql.gc_delete_range VALUES `)
 	require.Equal(t, ddl.BRInsertDeleteRangeSQLValue, `(%?, %?, %?, %?, %?)`)

